### PR TITLE
[full ci] Limit the number of soap requests in flight

### PR DESCRIPTION
**Pull request by caglar10ur**
Otherwise ESXi starts to kick us out and most of the portlayer code
cannot handle the error and fails to reconnect as of this moment.

This also protects ESXi from us.

Problem found while testing with following;

```bash
#!/bin/bash
COUNT=50

docker pull busybox

for i in $(seq $COUNT);
do
    docker create -t --name $i busybox /bin/sleep $i > /dev/null &
done

read -p "Press any key to continue"

for i in $(seq $COUNT);
do
    docker rm $i > /dev/null &
done
```

Helps #1723
Helps #5157

### CODE REVIEW

File: pkg/vsphere/tasks/waiter.go [hunk 1]
```diff
@@ -164,11 +201,11 @@ func logFault(op trace.Operation, fault types.BaseMethodFault) {
 }
 
 func logSoapFault(op trace.Operation, fault types.AnyType) {
-	op.Debugf("unexpected soap fault on task retry : %#v", fault)
+	op.Debugf("unexpected soap fault on task retry : %s", fault)
```
Comment (emlin) [inline pkg/vsphere/tasks/waiter.go]: somehow, print out all fields can help us understand what's wrong. Why change it back to %s, and curious why not for the above logFault?

File: pkg/vsphere/tasks/waiter.go [hunk 2]
```diff
@@ -39,6 +47,10 @@ type Task interface {
 	WaitForResult(ctx context.Context, s progress.Sinker) (*types.TaskInfo, error)
 }
 
+type temporary interface {
+	Temporary() bool
```
Comment (emlin) [inline pkg/vsphere/tasks/waiter.go]: is this for test purpose? didn't see test using it as well

File: pkg/vsphere/tasks/waiter.go [hunk 3]
```diff
@@ -60,6 +72,23 @@ func WaitForResult(ctx context.Context, f func(context.Context) (Task, error)) (
 	var info *types.TaskInfo
 	var backoffFactor int64 = 1
 
+	// initialize the limiter channel once
+	once.Do(func() {
+		inFlight = make(chan struct{}, maxInFlight)
+	})
+
+	// reserve the spot
+	inFlight <- struct{}{}
```
Comment (emlin) [inline pkg/vsphere/tasks/waiter.go]: This way will block the task directly without any message out.
Is thinking if do not want to have one more lock for available spots, might just get op before this, and log out message before and after get the spot.
So we could figure out the time gap to query one available task spot, to help debug.

**Comment by dougm** [comment]
I think this PR is fine to limit the number of `task.Wait` calls (and hence connections), as it does not block task from completing on the ESX side.

But the issue isn't so much the tasks in flight (as they run on the ESX side), but the number of property collectors we have waiting within the VCH for the ESX/VC tasks to complete.  Each call to `task.Wait` creates its own property collector and connection to wait for the result.

We could instead wait for N tasks on a single connection, rather than the current 1-to-1.  Need to think about / research the best way to do this.  How about we discuss tomorrow?

Related to hostd connection limitations: https://github.com/vmware/vic/issues/5621#issuecomment-316849333

And somewhat related, #5647, we could could avoid starting a new task (such as ReconfigVM) until a current task has completed.

File: pkg/vsphere/tasks/waiter.go [hunk 4]
```diff
@@ -60,6 +72,23 @@ func WaitForResult(ctx context.Context, f func(context.Context) (Task, error)) (
 	var info *types.TaskInfo
 	var backoffFactor int64 = 1
 
+	// initialize the limiter channel once
+	once.Do(func() {
```
Comment (hmahmood) [inline pkg/vsphere/tasks/waiter.go]: You could do this in `init()`.

File: pkg/vsphere/tasks/waiter.go [hunk 5]
```diff
@@ -60,6 +72,23 @@ func WaitForResult(ctx context.Context, f func(context.Context) (Task, error)) (
...
+	inFlight <- struct{}{}
+
+	// check whether we timed out/canceled
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+	}
+	// give the spot back
+	defer func() { <-inFlight }()
```
Comment (hmahmood) [inline pkg/vsphere/tasks/waiter.go]: This should be on line 82.

File: pkg/vsphere/tasks/waiter.go (hunk 1 above)
Comment (caglar10ur) [inline pkg/vsphere/tasks/waiter.go]: It was dumping the struct but not showing the error (instead it is showing the address of the err field)

File: pkg/vsphere/tasks/waiter.go (hunk 2 above)
Comment (caglar10ur) [inline pkg/vsphere/tasks/waiter.go]: it is used for the type assertion below

**Comment by caglar10ur** [comment]
@dougm agreed that we need to think more about this. I implemented this as a ugly/bad workaround for the bug in question but you are right this is not going to solve anything other than giving us some time

File: pkg/vsphere/tasks/waiter.go (hunk 5 above)
Comment (caglar10ur) [inline pkg/vsphere/tasks/waiter.go]: 👍  doing that but I think I won't be merging this until we discuss all the possible solutions first

**Comment by mhagen-vmware** [comment]
So, I am still a bit confused by this.. My understanding is that 6.0 has a task limit of 1024 and 6.5 upped that to 4096. On an empty system with us creating 75 containers do we really exceed 1024 task limit?  In my parallel PR #5621 when I actively monitored the system the task queue only got up to about 40 and the skd endpoint stopped responding. That tells me that the underlying problem is not necessarily the amount of calls we are making but what calls we are making?

**Comment by dougm** [comment]
@mhagen-vmware see my comment above "But the issue isn't so much the tasks in flight (as they run on the ESX side), but the number of property collectors we have waiting within the VCH for the ESX/VC tasks to complete. Each call to task.Wait creates its own property collector and connection to wait for the result."

In other words, we need to reduce/limit the number of *connections* we have open to ESX.   The number of *tasks* running on the ESX side isn't the issue here.

**Comment by mhagen-vmware** [comment]
Right, understood.. I saw that, but this PR will then only coincidentally solve this problem without directly solving the problem then right?  Or even worse this PR will just give us a potentially major performance hit without actually solving the underlying problem?

**Comment by caglar10ur** [comment]
I think the task limit and being able create that many tasks concurrently are different things. What I'm observing is happening only if you are doing things in parallel. As far my test goes we can't send more than 50-60 tasks in parallel against a single desktop PC that runs ESXi. This number does not include direct calls so I'm guessing we are hitting number of connections per session limit (which is 100 I believe)

**Comment by caglar10ur** [comment]
@mhagen-vmware @dougm @hickeng I'll schedule a short meeting for Monday to discuss this in detail and in the meantime I'll experiment a bit to have at least an idea what might be going on :)

**Comment by hmahmood** [comment]
@caglar10ur can you add me to the meeting too please?

**Comment by caglar10ur** [comment]
Change of plans, I started to limit number of  SOAP requests in-flight

**Comment by mhagen-vmware** [comment]
Parallel test build with this patch:
https://ci.vcna.io/vmware/vic/12602

File: pkg/vsphere/session/session.go [hunk 6]
```diff
@@ -86,6 +91,32 @@ type Session struct {
 	folders *object.DatacenterFolders
 }
 
+// RoundTripFunc alias
+type RoundTripFunc func(*http.Request) (*http.Response, error)
+
+// RoundTrip method
+func (rt RoundTripFunc) RoundTrip(r *http.Request) (*http.Response, error) {
```
Comment (hmahmood) [inline pkg/vsphere/session/session.go]: Is this used anywhere?

Comment (caglar10ur) [inline pkg/vsphere/session/session.go]: Yep, without it

```
building Portlayer API server...
# github.com/vmware/vic/pkg/vsphere/session
pkg/vsphere/session/session.go:119: cannot use RoundTripFunc(func literal) (type RoundTripFunc) as type http.RoundTripper in return argument:
        RoundTripFunc does not implement http.RoundTripper (missing RoundTrip method)
```

**Comment by caglar10ur** [comment]
@mhagen-vmware what is the setup in that test? Do we use a single esxi/single vch for multiple tests or we have multiple vch or esxi or both?

**Comment by mhagen-vmware** [comment]
@caglar10ur 1 ESX with up to 4 VCH during the run.  The ESX is a 4 CPU/8Core 32GB mem, SSD drives.

**Comment by caglar10ur** [comment]
bumped the default MaxIdleConnsPerHost from2 to maxInFlight this way we don't keep opening and closing connections instead reuse them (ESXi seems to be counting TIME_WAIT as a connection too so I believe this way we don't exceed the connection limit)

This is what I observe while creating 100 containers concurrently
```
Every 1.0s: govc host.esxcli network ip connection list | grep ":443"                                                                                                                                                        Tue Jul 25 22:50:26 2017

tcp      0        0        10.118.69.221:443                       10.118.67.228:55716  ESTABLISHED  66764     newreno  rhttpproxy-IO
tcp      0        0        10.118.69.221:443                       10.118.69.225:59916  ESTABLISHED  66764     newreno  rhttpproxy-IO
tcp      0        0        10.118.69.221:443                       10.118.69.225:59914  ESTABLISHED  66764     newreno  rhttpproxy-IO
tcp      0        0        10.118.69.221:443                       10.118.69.225:59912  ESTABLISHED  66764     newreno  rhttpproxy-IO
tcp      0        0        10.118.69.221:443                       10.118.69.225:59906  ESTABLISHED  66764     newreno  rhttpproxy-IO
tcp      0        0        10.118.69.221:443                       10.118.69.225:59904  ESTABLISHED  66763     newreno  rhttpproxy-IO
tcp      0        0        10.118.69.221:443                       10.118.69.225:59896  ESTABLISHED  66763     newreno  rhttpproxy-IO
tcp      0        0        10.118.69.221:443                       10.118.69.225:59892  ESTABLISHED  66764     newreno  rhttpproxy-IO
tcp      0        0        10.118.69.221:443                       10.118.69.225:59890  ESTABLISHED  66764     newreno  rhttpproxy-IO
tcp      0        0        10.118.69.221:443                       10.118.69.225:59888  ESTABLISHED  66764     newreno  rhttpproxy-IO
tcp      0        0        10.118.69.221:443                       10.118.69.225:59872  ESTABLISHED  66764     newreno  rhttpproxy-IO
tcp      0        0        10.118.69.221:443                       10.118.69.225:59862  ESTABLISHED  66764     newreno  rhttpproxy-IO
tcp      0        0        10.118.69.221:443                       10.118.69.225:59842  ESTABLISHED  66764     newreno  rhttpproxy-IO
tcp      0        0        10.118.69.221:443                       10.118.69.225:59834  ESTABLISHED  66764     newreno  rhttpproxy-IO
tcp      0        0        10.118.69.221:443                       10.118.69.225:59390  ESTABLISHED  66763     newreno  rhttpproxy-IO
tcp      0        0        10.118.69.221:443                       10.118.69.149:46886  ESTABLISHED  66763     newreno  rhttpproxy-IO
tcp      0        0        10.118.69.221:443                       10.118.67.228:50734  ESTABLISHED  66764     newreno  rhttpproxy-IO
tcp      0        0        0.0.0.0:443                             0.0.0.0:0            LISTEN       66758     newreno  rhttpproxy
tcp      0        0        :::443                                  :::0                 LISTEN       66758     newreno  rhttpproxy
```

**Comment by caglar10ur** [comment]
@mhagen-vmware do you mind trying with this version again?

**Comment by mhagen-vmware** [comment]
@caglar10ur https://ci.vcna.io/vmware/vic/12637

**Comment by caglar10ur** [comment]
just wanted to follow up, we are still working on this issue. Rate limiting the creation of vms uncovered another issue and currently debugging it with esxi/vcenter team.

**Comment by caglar10ur** [comment]
Just finished creating/starting/stopping/removing 128 VMs concurrently and rate limiter limited the operations 16 at a time as expected. I started to believe what we are observing in parallel tests are just slowness due to doing lots of things in parallel.

File: pkg/vsphere/session/session.go [hunk 7]
```diff
@@ -178,8 +211,19 @@ func (s *Session) Connect(ctx context.Context) (*Session, error) {
...
+			maxInFlight = i
+		}
+	}
+	// Limit the concurrenty of SOAP requests
+	if t, ok := soapClient.Transport.(*http.Transport); ok {
+		t.MaxIdleConnsPerHost = maxInFlight
+	}
+	soapClient.Transport = LimitConcurrency(soapClient.Transport, maxInFlight)
 
+	// TODO: option to set http.Client.Transport.TLSClientConfig.RootCAs
```
Comment (hickeng) [inline pkg/vsphere/session/session.go]: Missing RootCAs here seems like a significant omission as it could prevent us from accepting a legitimate vSphere.
Not that we have a means to provide a CA for vSphere specifically or to be added to the system pool.

File: pkg/vsphere/session/session.go [hunk 8]
```diff
@@ -282,6 +326,16 @@ func (s *Session) Populate(ctx context.Context) (*Session, error) {
...
 	}
 
+	if s.Datacenter != nil {
+		folders, err := s.Datacenter.Folders(ctx)
+		if err != nil {
+			errs = append(errs, fmt.Sprintf("Failure finding folders (%s): %s", s.DatacenterPath, err.Error()))
+		} else {
+			log.Debugf("Cached folders: %s", s.DatacenterPath)
+		}
+		s.VMFolder = folders.VmFolder
```
Comment (hickeng) [inline pkg/vsphere/session/session.go]: This is being used without being checked elsewhere - should be a hard error if it cannot be set.

File: pkg/vsphere/tasks/waiter.go [hunk 9]
```diff
@@ -152,6 +159,11 @@ func isRetryError(op trace.Operation, err error) bool {
 		if f, ok := err.(types.HasFault); ok {
 			logFault(op, f.Fault())
 		} else {
+			// retry the temporary errors
```
Comment (hickeng) [inline pkg/vsphere/tasks/waiter.go]: There is a DoNotRetry error used in `pkg/fetcher` and I'm sure similar logic in `pkg/retry`. I have a todo item to get retry and tasks package working together or unified. This is just informational.

File: pkg/vsphere/tasks/waiter.go (hunk 1 above)
Comment (hickeng) [inline pkg/vsphere/tasks/waiter.go]: The purposes of the tasks package were:
1. retry in the case of temporary errors, such as InProgress
2. operate in a queue aware fashion (such as throttling)
3. provide easy access to error types and messages

We can start adding code to extract the messages from the embedded faults and print that. Dumping the structs is massively verbose and hard to parse.

Also of note is that we get this even for _expected_ errors because we have no way of indicating whether an error is expected or not.

File: internal/govmomi/host_internal_config_manager.go [hunk 10]
```diff
@@ -0,0 +1,48 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
```
Comment (caglar10ur) [inline internal/govmomi/host_internal_config_manager.go]: Go documentation says it is possible ```any package in or under a directory named internal may be imported by packages rooted in the same subtree.``` but I not sure that is what we want as that naming dictates that only pkv/vsphere can import that directory.

File: infra/scripts/bash-helpers.sh [hunk 11]
```diff
@@ -90,15 +89,35 @@ vic-admin () {
...
+vic-tail-portlayer() {
+    unset keyarg
+    if [ -e $HOME/.ssh/authorized_keys ]; then
+        keyarg="--authorized-key=$HOME/.ssh/authorized_keys"
+    fi
+
+    out=$($(vic-path)/bin/vic-machine-"$OS" debug --target="$GOVC_URL" --compute-resource="$COMPUTE" --name=${VIC_NAME:-${USER}test} --enable-ssh "$keyarg" --rootpw=password --thumbprint=$THUMBPRINT $*)
+    host=$(echo $out | grep DOCKER_HOST | awk -F"DOCKER_HOST=" '{print $2}' | cut -d ":" -f1 | cut -d "=" -f2)
+
+    sshpass -ppassword ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root@${host} tail -f /var/log/vic/port-layer.log
```
Comment (emlin) [inline infra/scripts/bash-helpers.sh]: `-ppassword`
is that `-p password?`

File: lib/install/management/appliance.go [hunk 12]
```diff
@@ -202,6 +205,21 @@ func (d *Dispatcher) deleteVM(vm *vm.VirtualMachine, force bool) error {
...
+		if err := vm.Unregister(d.ctx); err != nil {
+			ok := false
+			if soap.IsSoapFault(err) {
+				// because internal
+				ok = soap.ToSoapFault(err).String == "vim.fault.ConcurrentAccess"
+			}
+			if !ok {
+				return fmt.Errorf("Failed to unregister VM %q: %s", vm.Reference(), err)
+			}
+			log.Debugf("Ignoring ConcurrentAccess error for %s", vm.Reference())
```
Comment (emlin) [inline lib/install/management/appliance.go]: what gonna happen if concurrent access error is returned? is that vm unregistered or still there

**Comment by caglar10ur** [comment]
This PR includes the other PRs for now. I'll keep rebasing to make this smaller once the dependencies go in to the repo.

**Comment by caglar10ur** [comment]
Removed all the other changes from this PR so this only does limiting but nothing else

**Comment by caglar10ur** [comment]
@emlin can you take a look at this PR again? I believe this now works without requiring the internal llpm calls

**Comment by emlin** [comment]
>@emlin can you take a look at this PR again? I believe this now works without requiring the internal llpm calls

Still didn't look yet. But you mean you'll not change to LLPM?

File: lib/install/management/appliance.go [hunk 13]
```diff
@@ -191,17 +191,30 @@ func (d *Dispatcher) deleteVM(vm *vm.VirtualMachine, force bool) error {
...
+	//removes the vm from vsphere, but detaches the disks first
 	_, err = vm.WaitForResult(d.ctx, func(ctx context.Context) (tasks.Task, error) {
 		return vm.DeleteExceptDisks(ctx)
 	})
 	if err != nil {
-		err = errors.Errorf("Failed to destroy VM %q: %s", vm.Reference(), err)
-		err2 := vm.Unregister(d.ctx)
-		if err2 != nil {
-			return errors.Errorf("%s then failed to unregister VM: %s", err, err2)
+		f, ok := err.(types.HasFault)
```
Comment (emlin) [inline lib/install/management/appliance.go]: hmm.. for vic-machine, if we reach here, we already powered off the VM, and plan to delete the vm even there is error happens, I guess.
If user wants to delete the VCH together with container, if we differentiate the error type here, user might have to rerun the delete, and does not care about the error anyway.

Comment (caglar10ur) [inline lib/install/management/appliance.go]: Good point, I'll fix the logic

File: lib/portlayer/exec/container.go [hunk 14]
```diff
@@ -596,6 +596,8 @@ func (c *Container) Remove(ctx context.Context, sess *session.Session) error {
 				log.Errorf("Error while attempting to unregister container VM: %s", err)
 				return err
 			}
+		case *types.ConcurrentAccess:
```
Comment (emlin) [inline lib/portlayer/exec/container.go]: is this error happens during vm delete or unregister? I think that's in unregister if you're using llpm, but not sure for here.
btw, if this error happens, do you need a return here?

File: pkg/vsphere/session/session.go [hunk 15]
```diff
@@ -180,8 +213,19 @@ func (s *Session) Connect(ctx context.Context) (*Session, error) {
 
 	soapClient.SetThumbprint(soapURL.Host, s.Thumbprint)
 
-	// TODO: option to set http.Client.Transport.TLSClientConfig.RootCAs
+	maxInFlight := defaultMaxInFlight
+	if e := os.Getenv("VIC_MAX_IN_FLIGHT"); e != "" {
+		if i, err := strconv.Atoi(e); err == nil {
+			maxInFlight = i
+		}
```
Comment (emlin) [inline pkg/vsphere/session/session.go]: might an else here, to log out the env setting is wrong.

File: pkg/vsphere/vm/vm.go [hunk 16]
```diff
@@ -262,10 +263,21 @@ func (vm *VirtualMachine) DeleteExceptDisks(ctx context.Context) (*object.Task,
...
 	disks := devices.SelectByType(&types.VirtualDisk{})
-
 	err = vm.RemoveDevice(ctx, true, disks...)
 	if err != nil {
-		return nil, err
+		switch f := err.(type) {
+		case task.Error:
+			switch f.Fault().(type) {
+			case *types.ConcurrentAccess:
+				log.Errorf("RemoveDevice failed for %s with ConcurrentAccess error. Ignoring it", vm)
```
Comment (emlin) [inline pkg/vsphere/vm/vm.go]: Just curious, does this mean someone else is removing device? Cause the device removing is critical for us, I have to confirm for this. 😄

File: lib/portlayer/exec/container.go (hunk 14 above)
Comment (caglar10ur) [inline lib/portlayer/exec/container.go]: I need to document those quirks, without llpm we are getting occasional ConcurrentAccess errors during DeleteExceptDisk, that's why we are ignoring this

File: pkg/vsphere/session/session.go (hunk 15 above)
Comment (caglar10ur) [inline pkg/vsphere/session/session.go]: VIC_MAX_IN_FLIGHT is useful only during development so I ignored this but let me know if you really want t that log line :)

File: lib/install/management/appliance.go [hunk 17]
```diff
@@ -191,17 +191,31 @@ func (d *Dispatcher) deleteVM(vm *vm.VirtualMachine, force bool) error {
...
-		if err2 != nil {
-			return errors.Errorf("%s then failed to unregister VM: %s", err, err2)
+		if f, ok := err.(types.HasFault); ok {
+			switch f.Fault().(type) {
+			case *types.ConcurrentAccess:
+				log.Warnf("DeleteExceptDisks failed with ConcurrentAccess error. Ignoring it.")
+				ignore = true
+			}
+		}
+		if !ignore {
```
Comment (emlin) [inline lib/install/management/appliance.go]: 👍
